### PR TITLE
fix(url-for-image): Rect overflows image bounds

### PR DESCRIPTION
**Pull request by sgulseth**
Fixes issue where the desired aspect ratio is a half where the rect built is overflowing the image bounds.

See https://sanity-io-land.slack.com/archives/C9Z7RC3V1/p1623324362116100 & https://codesandbox.io/s/boring-sound-8khon?file=/src/App.js

I would ideally like to add more tests to this with more cases.

An alternative fix is:
```js

cropRect = {
  left: Math.ceil(left),
  top: Math.round(top),
  width: Math.floor(width),
  height: Math.round(height),
}
```

As it rounds the width down and left crop up.

Edit:

Released on npm as `@sanity/image-url@0.140.23-canary.0`

### CODE REVIEW

**Review by judofyr — COMMENTED** [review]
I was wondering if an alternative approach could be to round earlier? If we make sure that `width` and `hotspotXCenter` are integers, then the rest of the calculation should be precise and then we can drop the rest of the `round`-calls:

```javascript
const width = Math.round(height * desiredAspectRatio)
// ...
const hotspotXCenter = Math.round((hotspot.right - hotspot.left) / 2 + hotspot.left)
```

We would then also need to round here though: https://github.com/sanity-io/image-url/blob/a2dc285ce09b6e1372f9a485f1ee0edfdfba45f0/src/urlForImage.ts#L59-L69

(Also `hotspotXCenter` can be also calculated with `(hotspot.right + hotspot.left) / 2` which is a bit shorter.)

**Comment by sgulseth** [comment]
Yes, that would work. However I think we would need to then assure that the given `CropSpec` is also rounded before applying it in `fit` since `urlForImage` accepts a custom `crop: CropSpec`[1] 🤔 I agree that dealing with integers are easier in `fit`.
Could we could do it without making it a breaking change? (Would be breaking because it it could move the crop by 1px if you specify your own crop)

[1]: https://github.com/sanity-io/image-url/blob/master/src/urlForImage.ts#L73-L74

**Comment by rexxars** [comment]
> Yes, that would work. And it might be a better long term solution as the output is `https://cdn.sanity.io/images/zp7mbokg/production/Tb9Ew8CXIwaY6R1kjMvI0uRR-3833x2555.jpg?w=600&h=400` instead of `https://cdn.sanity.io/images/zp7mbokg/production/Tb9Ew8CXIwaY6R1kjMvI0uRR-3833x2555.jpg?rect=1,0,3832,2555&w=600&h=400`

That would be great. It'll scale easier on the backend, too.

> Could we could do it without making it a breaking change? (Would be breaking because it it could move the crop by 1px if you specify your own crop).

I'm fine with bumping to 1.0.0 for this, to be honest. Takes us into the "logical semver range", too ;)

File: src/urlForImage.ts [hunk 1]
```diff
@@ -166,20 +166,20 @@ function fit(
     }
 
     cropRect = {
-      left: Math.round(left),
-      top: Math.round(top),
-      width: Math.round(width),
-      height: Math.round(height),
+      left: Math.max(0, left),
+      top: Math.max(0, top),
```
Comment (judofyr) [inline src/urlForImage.ts:170]: Should these `max` instead be applied when we define `left` and `top`? E.g:

```typescript
let left = Math.max(0, Math.round(hotspotXCenter - width / 2))
````

**Review by judofyr — COMMENTED** [review]
Approach looks good, but would be nice with more unit tests for the various edge cases.

**Review by judofyr — APPROVED** [review]
